### PR TITLE
feat: supoort Windows Terminal

### PR DESCRIPTION
**Pull request by TerakomariGandesblood**
Trying to support the preview version of Windows Terminal.
Known issues: https://github.com/benjajaja/ratatui-image/issues/37#issuecomment-2341352856

### CODE REVIEW

File: examples/async.rs [hunk 1]
```diff
@@ -42,7 +42,13 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     let backend = CrosstermBackend::new(stdout);
     let mut terminal = Terminal::new(backend)?;
 
+    #[cfg(all(feature = "rustix", unix))]
     let mut picker = Picker::from_termios()?;
+    #[cfg(not(all(feature = "rustix", unix)))]
+    let mut picker = {
+        let font_size = (8, 16);
+        Picker::new(font_size)
+    };
```
Comment (j4james) [inline examples/async.rs:51]: If this is used to calculate the amount of space an image will occupy in character cells, it should be 10x20 for Windows Terminal (that's the cell size of the original VT340 hardware terminal).

You can also query the cell size with the `CSI 16 t` escape sequence.

File: src/picker.rs [hunk 2]
```diff
@@ -252,6 +261,46 @@ fn guess_protocol() -> (ProtocolType, bool) {
     (ProtocolType::Halfblocks, is_tmux)
 }
 
+// Since only the latest preview version of Windows Terminal supports Sixels,
+// it is necessary to get the version and determine whether it supports Sixels or not.
+#[cfg(windows)]
+fn support_sixels() -> Result<bool> {
```
Comment (j4james) [inline src/picker.rs:267]: You should be able to determine sixel support from the device attributes query, as is already done here:

https://github.com/benjajaja/ratatui-image/blob/e630e680a186fc0c44c43be0e6966889094e525e/src/picker.rs#L344-L346

That way you'll be able to support sixels in other terminals as well.

**Review by benjajaja — CHANGES_REQUESTED** [review]
@j4james [pointed out](https://github.com/microsoft/terminal/issues/8581#issuecomment-2395553315) elsewhere that windows terminal effectively has a way to [query font size](https://github.com/microsoft/terminal/pull/17504)!

So we can query the font size and sixel support without having to rely on guessing by processes and version files and so on. But to query, we currently do a bunch of things.

On POSIX, we put the terminal into canonical mode, disable echoing, and use fcntl to set the terminal in non-blocking mode.

On windows, I think (with the help of ChatGPT) that we should use SetConsoleMode to disable ENABLE_ECHO_INPUT and ENABLE_LINE_INPUT, and use PeekConsoleInput to read input without blocking indefinitely.

My plan of refactor would be:
1. Try out `CSI 16 t` (or 14) to get the font-size without `tcgetwinsize` first to test the waters.
2. Let `query_device_attrs` take the query as param, and let it return the response string.
3. Put the response->protocol parsing somewhere else, e.g. just inside `guess_protocol`.
4. Provide a windows version of `query_device_attrs` with the above mentioned Windows API calls instead of the POSIX API calls.
5. Consider removing `from_termios()` and just have one method/constructor that queries the terminal for everything :heart_eyes:

**Comment by TerakomariGandesblood** [comment]
Sounds good. I'm not familiar enough with terminals and Windows. It will take a while to work on it.

**Comment by benjajaja** [comment]
I can take care of steps 1-3.

**Comment by benjajaja** [comment]
@TerakomariGandesblood I did some changes to `Picker` on `master`. You should try it out, it should just work with the terminal of WSL if I understand correctly. For non-WSL (so only halfblocks but proper alignement) it may also be possible to translate `query_stdio_capabilities` to the Windows Console API.

**Comment by TerakomariGandesblood** [comment]
Great job, I'll check it out in a few days

**Comment by TerakomariGandesblood** [comment]
Already implemented this feature in other PR, so close this PR.

**Pull request closed without merging**